### ISSUE TITLE
distsql: query hung after cancellation

### ISSUE THREAD

**Issue by asubiotto** [cockroachdb/cockroach#19951]
Ran a long-running read query (i.e. `SELECT * FROM tpch.lineitem ORDER BY l_extendedprice`) on a two-node cluster and canceled. The query has been hung and an initial glance seems to indicate that the [noop processor](https://cockroachdb.github.io/distsqlplan/decode.html?eJzUkUFLw0AQhe_-CpmrI3SStGpOc-2llepNcojZoSykmbC7AaX0v0uyh9rSxApePO7MfO899u2hUSOrcice8jcgQEigQGidVuK9un4cj5bmA_IZgm3aLvTjAqFSJ5DvIdhQC-TwWr7XspHSiAMEI6G09SDcOrsr3SfXthEbZAcI6y7kt0zICXKKnCHPkRfID8iPyE_INEMmQqYEmVJkypBpjkwLKA4I2oVjEB_KrUBOB7w-7Iu6cJ6TF3ej4smo-FFTnREn5pLohQQrvdf25HbMOz3xpv_Uwg9h_66FC-Ib8a02Xq7641lfkZitxEq9dq6SZ6fVYBOf64EbBkZ8iFuKj2UTV33A7zBNwskJTOdwMgnPp53TSTibhrNfxS4ON18BAAD__y6EXuo=) on the first node has been cleaned up while the second node's sorter is still trying to push rows.

What leads me to believe this is that the second node is stuck on:
```
goroutine 583 [chan send, 9 minutes]:
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*RowChannel).Push(0xc42146d8c0, 0xc427096000, 0x10, 0x10, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:337 +0x1ba
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcOutputHelper).EmitRow(0xc42018ad80, 0x6845140, 0xc4207beb10, 0xc423996300, 0x10, 0x10, 0x0, 0x0, 0x0)
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:302 +0x2be
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*sortAllStrategy).executeImpl(0xc4203742e0, 0x6845140, 0xc4207beb10, 0xc42018ad80, 0x6846080, 0xc4200a2d80, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sorterstrategy.go:150 +0x2bc
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*sortAllStrategy).Execute(0xc4203742e0, 0x6845140, 0xc4207beb10, 0xc42018ad80, 0x0, 0x0)
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sorterstrategy.go:105 +0x537
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*sorter).Run(0xc42018ad80, 0x6845140, 0xc4207beb10, 0xc4207529b0)
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sorter.go:145 +0x5d8
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:400 +0x3d7
```

And the first node is stuck processing this (trying to push to the noop processor):
```
goroutine 157923 [chan send, 9 minutes]:
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*RowChannel).Push(0xc4214fdc80, 0xc424814a00, 0x10, 0x10, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:337 +0x1ba
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.processInboundStreamHelper(0x6845140, 0xc4214fdec0, 0x6851020, 0xc42030f050, 0x0, 0x6832d00, 0xc4214fdc80, 0xc4205b6a00, 0xc42f45cb00, 0x0)
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/inbound.go:118 +0x2dc
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.ProcessInboundStream(0x6845140, 0xc4214fdec0, 0x6851020, 0xc42030f050, 0xc42fb2ec60, 0x6832d00, 0xc4214fdc80, 0xc4205b6a00, 0x2540be400, 0xc4205b6a00)
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/inbound.go:39 +0x9b
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ServerImpl).flowStreamInt(0xc4200ceb00, 0x6845140, 0xc4214fdec0, 0x6851020, 0xc42030f050, 0x0, 0x0)
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/server.go:401 +0x428
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ServerImpl).FlowStream(0xc4200ceb00, 0x6851020, 0xc42030f050, 0x683e280, 0xc4200ceb00)
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/server.go:407 +0x91
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun._DistSQL_FlowStream_Handler(0x5a657c0, 0xc4200ceb00, 0x684e620, 0xc420220400, 0x4014aa8, 0x20)
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/api.pb.go:367 +0xb2
github.com/cockroachdb/cockroach/pkg/rpc.NewServerWithInterceptor.func2(0x5a657c0, 0xc4200ceb00, 0x684e620, 0xc420220400, 0xc42123e960, 0x5b97f20, 0x4040441, 0x261d8)
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:190 +0x103
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).processStreamingRPC(0xc42009ec80, 0x68506c0, 0xc424a9c160, 0xc4297de400, 0xc4202e7620, 0x6d27fa0, 0x0, 0x0, 0x0)
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:940 +0x2e8
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleStream(0xc42009ec80, 0x68506c0, 0xc424a9c160, 0xc4297de400, 0x0)
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1027 +0x14c1
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc423a87080, 0xc42009ec80, 0x68506c0, 0xc424a9c160, 0xc4297de400)
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:572 +0x9f
created by github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
	/Users/asubiotto/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:570 +0xa1
```

With no indication of any processors' presence on the first (gateway) node.

A user has run into this issue in #19797, but we have not yet established if the cause is the same.

**Comment by asubiotto** [comment]
Here is me tracking down the bug:

On the gateway node, we set up a noop processor that reads from an ordered synchronizer that reads from a bunch of rowchannels to which data is pushed. Instrumenting the processor, it seems that when draining properly, we hit the case that the `emitHelper` returns false and does the "correct" thing (i.e. whatever it is to make the query not hang). So when we don't drain properly, we're getting back a nil row and no metadata from the ordered synchronizer, so what is going on there?

Essentially the input synchronizer is set to a drained state, which returns a nil row and no metadata, causing the noop to believe that the input synchronizer has no more rows to offer. In this case, the noop processor simply quits, closing the outputs but doing nothing to the inputs.

However, closing the inputs at this stage in the noop processor is actually a noop (per ConsumerClosed semantics, it is a noop if called after Next), so that implies that proper closing of the inputs is the responsibility of the ordered synchronizer. In the ordered synchronizer, there is a method called `consumeMetadata` which apparently consumes everything from the inputs, forwarding only metadata. However, from the goroutine dump, we can see that a row is still trying to be pushed to the RowChannel (which then pushes this onto the ordered synchronizer), so why is this not working as expected?

Instrumenting `consumeMetadata` it seems that no streams are consumed, and with further instrumentation, we can see that there is a discrepancy between the size of the heap (0), which keeps tracks of the "current" row from each stream and the number of input streams (2). The input synchronizer is indeed set up with two streams, so there seems to be some fault in the initialization.

Looking at the initialization, there seems to be an error initializing the heap ("query execution canceled") which is an expected error. However, we then call `ConsumerClosed` which will set our state to `draining` (i.e. we lose our `not initialized` state), so on the next call to `Next` (remember that this error is only forwarded to the consumer by the noop processor), we essentially move on to the normal draining code which iterates over the heap. We iterate over the heap rather than the streams because the heap can be reduced in size when you've finished consuming rows from an input but not another one (so you don't care about finished streams).

So going back a bit, the fact that `consumeMetadata` doesn't consume streams is fine, because it will return with a context canceled error anyway. The issue is that when draining, we should check whether the heap has been initialized, because otherwise the assumption that a heap size of 0 implies that all streams have been consumed is false. We should check when draining whether we are transitioning from an uninitialized state, and in that case, close our inputs (which will unblock the `RowChannel`).

That's right, and it turns out that that functionality has, in fact, been implemented (check out `consumerStatusChanged`) with the only caveat that in `ConsumerDone`, we set our state to `draining` *before* calling `consumerStatusChanged`, so when transitioning from uninitialized to draining, we never actually call `ConsumerClosed` on the sources.

We can fix this by switching the state change to after `consumerStatusChanged`. To fix other scenarios which might suffer from the same issue, let's just change `consumerStatusChanged` to do the state transition for us.

And as soon as you think you have the answer, it turns out that wasn't it. Let's verify that we're calling consumerdone on the input and that that is resulting in closing the blocking channel.

**Comment by asubiotto** [comment]
On the gateway node, we set up a noop processor that reads from an ordered synchronizer that reads from a bunch of rowchannels to which data is pushed. Instrumenting the processor, it seems that when draining properly, we hit the case that the `emitHelper` returns false and does the "correct" thing (i.e. whatever it is to make the query not hang). So when we don't drain properly, we're getting back a nil row and no metadata from the ordered synchronizer, so what is going on there?

Essentially the input synchronizer is set to a drained state, which returns a nil row and no metadata, causing the noop to believe that the input synchronizer has no more rows to offer. In this case, the noop processor simply quits, closing the outputs but doing nothing to the inputs.

However, closing the inputs at this stage in the noop processor is actually a noop (per ConsumerClosed semantics, it is a noop if called after Next), so that implies that proper closing of the inputs is the responsibility of the ordered synchronizer. In the ordered synchronizer, there is a method called `consumeMetadata` which apparently consumes everything from the inputs, forwarding only metadata. However, from the goroutine dump, we can see that a row is still trying to be pushed to the RowChannel (which then pushes this onto the ordered synchronizer), so why is this not working as expected?

Instrumenting `consumeMetadata` it seems that no streams are consumed, and with further instrumentation, we can see that there is a discrepancy between the size of the heap (0), which keeps tracks of the "current" row from each stream and the number of input streams (2). The input synchronizer is indeed set up with two streams, so there seems to be some fault in the initialization.

Looking at the initialization, there seems to be an error initializing the heap ("query execution canceled") which is an expected error. However, we then call `ConsumerClosed` which will set our state to `draining` (i.e. we lose our `not initialized` state), so on the next call to `Next` (remember that this error is only forwarded to the consumer by the noop processor), we essentially move on to the normal draining code which iterates over the heap. We iterate over the heap rather than the streams because the heap can be reduced in size when you've finished consuming rows from an input but not another one (so you don't care about finished streams).

So going back a bit, the fact that `consumeMetadata` doesn't consume streams is fine, because it will return with a context canceled error anyway. The issue is that when draining, we should check whether the heap has been initialized, because otherwise the assumption that a heap size of 0 implies that all streams have been consumed is false. We should check when draining whether we are transitioning from an uninitialized state, and in that case, close our inputs (which will unblock the `RowChannel`).

That's right, and it turns out that that functionality has, in fact, been implemented (check out `consumerStatusChanged`) with the only caveat that in `ConsumerDone`, we set our state to `draining` *before* calling `consumerStatusChanged`, so when transitioning from uninitialized to draining, we never actually call `ConsumerDone` on the sources.

We can fix this by switching the state change to after `consumerStatusChanged`. To fix other scenarios which might suffer from the same issue, let's just change `consumerStatusChanged` to do the state transition for us.

And as soon as you think you have the answer, it turns out that wasn't quite it. However, the next step is verifying that we are actually calling `ConsumerDone` which turns out to be correct. The issue is, following this call down to the `RowChannel`, is that only `ConsumerClosed` unblocks the channel. Looking at the `RowSource` interface for the semantics, it seems that `ConsumerDone` means "I'm done with rows but will keep on consuming metadata", and that `ConsumerClosed` means "I'm done with everything".

Taking a step back, the draining flow is: hit something that puts you into draining mode, and then take care of draining in the next call to `Next()`. The issue is that we treat uninitialized streams the same as finished streams, and because neither are present in the heap, our `drainSources` code will miss these uninitialized streams. To stay with the general idea that initialized streams should be consumed, we leave the code as is, except add a call to `ConsumerClosed` after the usual draining code. This will be a noop on streams that have already had `ConsumerClosed` called on them and will correctly close uninitialized streams. Since calling `ConsumerClosed` sets our state to `drainBuffered` and we will only ever call `ConsumerClosed` when in state `draining`, and there exists no transition from `drainBuffered` to draining, we will never double-close a stream. For this to actually take effect, we get rid of the `isInitialized` check in `consumerStatusChange` (that calls the given `RowSource` function on the streams only if not initialized) to always operate on the sources rather than the heap. That distinction was made, according to a comment, because of some potential performance gains. However, I think we're a long way from caring about this.
